src/service/notification_service/email_sendgrid.rs

```diff
@@ -0,0 +1,142 @@
+use super::{email::EmailMessage, NotificationEmailTransportApi, Result};
+use async_trait::async_trait;
+use rocket::serde::json;
+use serde::Serialize;
+use serde_json::Value;
+
+#[derive(Debug, Clone)]
+pub struct SendgridConfig {
+    api_key: String,
+    url: String,
+}
+
+pub struct SendgridTransport {
+    config: SendgridConfig,
+    client: reqwest::Client,
+}
+
+impl SendgridTransport {
+    /// Creates a new instance of the SendgridTransport.
+    #[allow(dead_code)]
+    pub fn new(config: &SendgridConfig) -> Self {
+        let client = reqwest::Client::new();
+        Self {
+            config: config.to_owned(),
+            client,
+        }
+    }
+
+    async fn send_http_request(&self, message: SendgridMessage) -> Result<()> {
+        let url = format!("{}/v3/mail/send", self.config.url);
+        let request = self
+            .client
+            .post(url)
+            .json(&message)
+            .bearer_auth(&self.config.api_key)
+            .header("Content-Type", "application/json");
+        let _ = request.send().await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl NotificationEmailTransportApi for SendgridTransport {
+    async fn send(&self, message: EmailMessage) -> Result<()> {
+        self.send_http_request(message.try_into()?).await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+struct SendgridMessage {
+    // This is very dynamic so I just use json value here
```

Comment:
Maybe a line, or two of docs on what `personalizations` are here would be nice. I never used Sendgrid and have no idea what this might be :sweat_smile: the other fields seem self-explanatory.

Reply:
You can add sort of anything here (headers, templates, variables, ...). When we know what Sendgrid features we actually want to use I will add the required structs that can be serialized into this list. Actually for most of these options I also have no idea what they are for :-)